File: krishang/src/helios/contracts/artifact.py
import hashlib
import json
from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import Field

from helios.clock import now_utc
from helios.ids import new_id
from .common import WireModel


class ArtifactType(StrEnum):
    PLAN = "plan"
    CLASSIFICATION = "classification"
    DUP_REPORT = "dup_report"
    RESEARCH = "research"
    REPRO_REPORT = "repro_report"
    PATCH = "patch"
    TEST_RESULT = "test_result"
    SECURITY_REPORT = "security_report"
    REVIEW_NOTES = "review_notes"
    DRAFT_REPLY = "draft_reply"
    CRITIC_VERDICT = "critic_verdict"
    BLOCKED = "blocked"
    ESCALATION = "escalation"
    RELEASE_DRAFT = "release_draft"
    REQUIREMENTS_SPEC = "requirements_spec"
    ARCHITECTURE_DECISION = "architecture_decision"
    IMPLEMENTATION_PLAN = "implementation_plan"
    BUILD_MANIFEST = "build_manifest"
    PACKAGE_RESULT = "package_result"
    DEPLOYMENT_DRAFT = "deployment_draft"
    REPOSITORY_INVENTORY = "repository_inventory"
    DEPENDENCY_INVENTORY = "dependency_inventory"
    SBOM = "sbom"
    SECRET_FINDING = "secret_finding"
    VULNERABILITY_FINDING = "vulnerability_finding"
    THREAT_MODEL = "threat_model"
    REMEDIATION_PLAN = "remediation_plan"
    SARIF_REPORT = "sarif_report"
    WRITEBACK_INTENT = "writeback_intent"


class Artifact(WireModel):
    schema_version: str = "1.0"
    artifact_id: str = Field(default_factory=lambda: new_id("artifact"))
    task_id: str
    run_id: str
    artifact_type: ArtifactType
    producer: str
    producer_version: str = "1.0"
    upstream_artifact_ids: list[str] = Field(default_factory=list)
    policy_ids: list[str] = Field(default_factory=list)
    content: dict[str, Any]
    content_hash: str
    created_at: datetime = Field(default_factory=now_utc)

    @classmethod
    def create(cls, *, content: dict[str, Any], **values: Any) -> "Artifact":
        encoded = json.dumps(content, sort_keys=True, separators=(",", ":"), default=str).encode()
        return cls(content=content, content_hash=hashlib.sha256(encoded).hexdigest(), **values)

File: Member 2/packages/contracts/src/artifact.ts
import { z } from "zod";
import { BodyText, EpochMs, RetentionClass, Sha256, ShortText, externalObject } from "./common";
import { ArtifactId, RunId, TaskId } from "./ids";

export const ArtifactType = z.enum([
  "plan", "classification", "dup_report", "research", "repro_report", "patch", "test_result", "security_report", "review_notes",
  "draft_reply", "critic_verdict", "blocked", "escalation", "release_draft", "requirements_spec", "architecture_decision",
  "implementation_plan", "build_manifest", "package_result", "deployment_draft", "repository_inventory", "dependency_inventory",
  "sbom", "secret_finding", "vulnerability_finding", "threat_model", "remediation_plan", "sarif_report",
]);

export const ArtifactProducer = z.object({ name: ShortText, version: ShortText }).strict();
export const Artifact = externalObject({
  artifactId: ArtifactId,
  taskId: TaskId,
  runId: RunId,
  nodeId: z.string().regex(/^node_[A-Za-z0-9_-]{1,80}$/),
  type: ArtifactType,
  producer: ArtifactProducer,
  upstreamArtifactIds: z.array(ArtifactId).max(200),
  policyRuleIds: z.array(z.string().min(1).max(128)).max(100),
  contentHash: Sha256,
  content: BodyText,
  retentionClass: RetentionClass,
  createdAt: EpochMs,
});

export const CriticVerdict = z.object({
  verdict: z.enum(["pass", "revise", "blocked"]),
  reviewedArtifactId: ArtifactId,
  reviewedContentHash: Sha256,
  producerAgent: ShortText,
  criticAgent: ShortText,
  criteria: z.array(z.object({ criterion: z.string().min(1).max(1_024), passed: z.boolean(), note: z.string().max(2_048) }).strict()).min(1).max(100),
  targetNodeId: z.string().max(96).optional(),
  rejectionFingerprint: Sha256.optional(),
}).strict().refine((value) => value.producerAgent !== value.criticAgent, "critic must be independent");

export type ArtifactType = z.infer<typeof ArtifactType>;
export type Artifact = z.infer<typeof Artifact>;
export type CriticVerdict = z.infer<typeof CriticVerdict>;
